components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/OAuthAppTenantResolverValve.java

```diff
@@ -0,0 +1,157 @@
+
+/**
+ * This valve is used to resolve the tenant domain of the OAuth application.
+ *
+ * When tenant qualified urls are not enabled, we need to set the tenant domain of the oauth app to the thread
+ * local. This is because with the client id tenant uniqueness improvement, DAO layer requires the tenant domain
+ * and client id to retrieve an app and the tenant is not available in the request path. Note that when tenant
+ * qualified urls are disabled, client id is unique across the server.
+ */
+public class OAuthAppTenantResolverValve extends ValveBase {
+
+    private static final Log log = LogFactory.getLog(OAuthAppTenantResolverValve.class);
+    private static String oAuthServerBaseURL = null;
+    private static String oAuth2ServerBaseURL = null;
+
+    @Override
+    public void invoke(Request request, Response response) throws IOException, ServletException {
+
+        try {
+            if (isOAuthRequest(request)) {
+                String appTenant = "";
+                String clientId = request.getParameter(CLIENT_ID);
+
+                if (StringUtils.isEmpty(clientId) && isOAuth10ARequest(request)) {
+                    clientId = request.getParameter(OAUTH_CONSUMER_KEY);
+                }
+
+                // If empty, try to get the client id from the authorization header.
+                if (StringUtils.isEmpty(clientId)) {
+                    if (OAuth2Util.isBasicAuthorizationHeaderExists(request)) {
+                        try {
+                            String[] credentials = OAuth2Util.extractCredentialsFromAuthzHeader(request);
+                            if (credentials != null) {
+                                clientId = credentials[0];
+                            }
+                        } catch (OAuthClientAuthnException e) {
+                            if (log.isDebugEnabled()) {
+                                log.debug("Error while extracting credentials from authorization header.", e);
+                            }
+                        }
+                    }
+                }
+
+                if (StringUtils.isNotEmpty(clientId)) {
+                    try {
+                        OAuthAppDO oAuthAppDO = OAuth2Util.getAppInformationByClientIdOnly(clientId);
+                        appTenant = OAuth2Util.getTenantDomainOfOauthApp(oAuthAppDO);
+                    } catch (IdentityOAuth2Exception e) {
+                        log.error("Error while getting oauth app for client Id.", e);
+                    } catch (InvalidOAuthClientException e) {
+                        if (log.isDebugEnabled()) {
+                            log.debug("Error while getting oauth app for client Id: " + clientId, e);
+                        }
+                    }
+                }
+
+                // Set the tenant name to thread local properties.
+                if (StringUtils.isNotEmpty(appTenant)) {
```

Comment:
Don't we need to break the flow if the appTenant is empty?

Reply:
Resolving as per the offline chat and https://github.com/wso2-extensions/identity-carbon-auth-rest/pull/240#discussion_r1338416845